_posts/2021-05-06-broker-tuning.md

```diff
@@ -0,0 +1,441 @@
+# ...
+```
+
+In Strimzi, you configure these settings through the `config` property of the `Kafka` custom resource.
+In this post, we suggest what else you might add to optimize your Kafka brokers.
+
+> Some properties are [managed directly by Strimzi](https://strimzi.io/docs/operators/latest/using.html#property-kafka-config-reference), such as `broker.id`. These properties are ignored if they are added to the `config` specification.
+
+### Replicating topics for high availability
+
+When you configure topics, the number of partitions, minimum number of in-sync replicas and partition replication factor are typically set at the topic level.
+You can use [Strimzi's `KafkaTopic`](https://strimzi.io/docs/operators/latest/using.html#proc-configuring-kafka-topic-str) to do this.
+You can also set these properties at the broker level too. In which case, the defaults are applied to topics that do not have these properties set explicitly, including automatically-created topics.
+
+```properties
+# ...
+num.partitions=1
+default.replication.factor=3
+min.insync.replicas=2
+# ...
+```
+
+High availability environments require a replication factor of at least 3 for topics and a minimum number of in-sync replicas as 1 less than the replication factor.
+For increased data durability, set `min.insync.replicas` in your *topic* configuration and message delivery acknowledgments using `acks=all` in your *producer* configuration.
+
+Use the `replica.fetch.max.bytes` property to set a maximum size for the messages fetched by each _follower_ that replicates the data from a _leader_ partition.
+Base the value on the average message size and throughput. When considering the total memory allocation required for read/write buffering, the memory available must also be able to accommodate the maximum replicated message size when multiplied by all followers. The size must also be greater than `message.max.bytes` so that all messages can be replicated.
+
+```properties
+# ...
+replica.fetch.max.bytes=1048576
+# ...
+```
+
+The importance of Kafka's topic replication mechanism cannot be overstated.
+Topic replication is central to Kafka's reliability and data durability. Using replication, a failed broker can recover from its in-sync replicas.
+We go into more detail about leaders, followers and in-sync replicas with [partition rebalancing for availability](#partition-rebalancing-for-availability).
+
+The `auto.create.topics.enable` property to create topics automatically is usually disabled. Kafka users tend to prefer applying more control over topic creation.
+If you do use automatic topic creation, set `num.partitions` to equal the number of brokers in the cluster so that writes are distributed.
+The `delete.topic.enable` property to allow topics to be deleted is usually disabled too. This time you're making sure that data is not accidentally lost, although you can temporarily enable the property to delete topics if circumstances demand it.
```

Comment:
Changed to:

> The `delete.topic.enable` property is enabled by default to allow topics to be deleted and is usually disabled too.

Reply:
Same as above ...

> TBH, I do not think this makes it clear to me ... the _but it's usually disabled_ makes no sense if it is enabled by default because it will not disable it self. I think it should say for example _but users usually disable it_ to make it clear that it is the user's responsibility to disable it.

Reply:
Changed to:

> The `delete.topic.enable` property is enabled by default to allow topics to be deleted. Kafka users normally disable this property too. This time you're making sure that data is not accidentally lost, although you can temporarily enable the property to delete topics if circumstances demand it.